File: composition/Scala/composition-pipeline/pipeline/src/test/scala/PipelineRunUntilFirstFaultTests.scala
import common.Pipeline._
import DomainModel._
import org.scalatest.FunSpec

class PipelineRunUntilFirstFaultTests extends FunSpec {

  val NAME_STUB = "John Kidd"
  val EMAIL_ADDRESS_STUB = "john@test.com"

  describe("Pipeline") {
     describe("runUntilFirstFault") {
      it("should return that we have no errors") {
        // arrange
        val personStub = Person(NAME_STUB, EMAIL_ADDRESS_STUB)
        val expected = ""

        // act
        val result = runUntilFirstFault(personStub, validateName, validateEmailAddress)
        val actual = result.error

        // assert
        assert(expected == actual)
      }

      it("should return that we have an invalid name") {
        // arrange
        val personStub = Person(emailAddress = EMAIL_ADDRESS_STUB)
        val expected = INVALID_NAME_MESSAGE

        // act
        val result = runUntilFirstFault(personStub, validateName, validateEmailAddress, updateName(NAME_STUB), updateEmailAddress(EMAIL_ADDRESS_STUB))
        val actual = result.error

        // assert
        assert(expected == actual)
      }

      it("should return that we the email address has an incorrect format") {
        // arrange
        val personStub = Person(name = NAME_STUB)
        val expected = EMAIL_ADDRESS_FORMAT_IS_INCORRECT_MESSAGE

        // act
        val result = runUntilFirstFault(personStub, validateName, validateEmailAddress, updateName(NAME_STUB), updateEmailAddress(EMAIL_ADDRESS_STUB))
        val actual = result.error

        // assert
        assert(expected == actual)
      }

      it("should return that we have an invalid name and break") {
        // arrange
        val personStub = Person()
        val expected = INVALID_NAME_MESSAGE

        // act
        val result = runUntilFirstFault(personStub, validateName, validateEmailAddress, updateName(NAME_STUB), updateEmailAddress(EMAIL_ADDRESS_STUB))
        val actual = result.error

        // assert
        assert(expected == actual)
      }
    }
  }
}

File: composition/C#/CompositionPipeline.Tests/PipelineRunUntilFirstFaultTests.cs
using System;
using NUnit.Framework;
using static CompositionPipeline.Common.Pipeline;
using static CompositionPipeline.DomainModel;
using static NUnit.Framework.Assert;
using System.Collections.Generic;
using CompositionPipeline.Common;

namespace CompositionPipeline.Tests
{
	[TestFixture]
	public class PipelineRunUntilFirstFaultTests
	{
		const string NAME_STUB = "John Kidd";
		const string EMAIL_ADDRESS_STUB = "john@test.com";

		[Test]
		public void should_return_that_we_have_no_errors() {
			// arrange
			var personStub = new Person(NAME_STUB, EMAIL_ADDRESS_STUB);
			var expected = "";
            var funcs = new List<Func<Person, Result<Person>>> { ValidateName, ValidateEmailAddress };

            // act
            var result = RunUntilFirstFault(personStub, funcs);
			var actual = result.Error;

			// assert
			AreEqual(expected, actual);
		}

		[Test]
		public void should_return_that_email_address_has_an_incorrect_format() {
			// arrange
			const string INCORRECTLY_FORMED_EMAIL_ADDRESS_STUB = "john.com";
			var personStub = new Person(NAME_STUB, INCORRECTLY_FORMED_EMAIL_ADDRESS_STUB);
			var expected = EMAIL_ADDRESS_FORMAT_IS_INCORRECT_MESSAGE;
            var funcs = new List<Func<Person, Result<Person>>> { ValidateName, ValidateEmailAddress };

            // act
            var result = RunUntilFirstFault(personStub, funcs);
			var actual = result.Error;

			// assert
			AreEqual(expected, actual);
		}
	}
}
